### PR TITLE
Slice the last matmull in stateful llm pipeline

### PR DESCRIPTION
**Pull request by olpipi**
Ticket: CVS-154175

### CODE REVIEW

File: src/cpp/src/llm_pipeline.cpp [hunk 1]
```diff
@@ -65,6 +68,22 @@ std::pair<EncodedResults, int32_t> beam_search(
 );
 
 class StatefulLLMPipeline final : public LLMPipelineImplBase {
+private:
+    void slice_matmul_statefull_model(std::shared_ptr<ov::Model> model) {
+        auto last_node = model->output(0).get_node()->input_value(0).get_node();
+        if (auto matmul = dynamic_cast<ov::op::v0::MatMul*>(last_node)) {
+            auto shape = matmul->input(0).get_partial_shape();
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline.cpp:75]: It seems you don't access `StatefulLLMPipeline`'s state. `slice_matmul_statefull_model()` should be standalone function.

File: src/cpp/src/llm_pipeline.cpp [hunk 2]
```diff
@@ -65,6 +68,22 @@ std::pair<EncodedResults, int32_t> beam_search(
...
+            if (shape.rank().get_length() == 3 && shape[1] != 1) {
+                auto start = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-1});
+                auto stop = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-2});
+                auto step = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-1});
+                auto axis = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{1});
+                auto slice = std::make_shared<ov::op::v8::Slice>(matmul->input_value(0), start, stop, step, axis);
+                matmul->input(0).replace_source_output(slice);
+            }
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline.cpp:83]: I think else brunch must throw so we don't need to guess why no improvement is observed.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/llm_pipeline.cpp (hunk 2 above)
Comment (ilya-lavrenov) [reply src/cpp/src/llm_pipeline.cpp:83]: or we can cast and add OPENVINO_ASSERT that casted value is not nullptr

File: src/cpp/src/llm_pipeline.cpp [hunk 3]
```diff
@@ -65,6 +68,22 @@ std::pair<EncodedResults, int32_t> beam_search(
 );
 
 class StatefulLLMPipeline final : public LLMPipelineImplBase {
+private:
+    void slice_matmul_statefull_model(std::shared_ptr<ov::Model> model) {
+        auto last_node = model->output(0).get_node()->input_value(0).get_node();
+        if (auto matmul = dynamic_cast<ov::op::v0::MatMul*>(last_node)) {
+            auto shape = matmul->input(0).get_partial_shape();
+            if (shape.rank().get_length() == 3 && shape[1] != 1) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline.cpp:76]: why do we need a check for shape[1] (sequence length) ? I should be -1 by default as model is exported as dynamic one

Comment (itikhono) [reply src/cpp/src/llm_pipeline.cpp:76]: this check will throw exception if rank or shape[1] dim are dynamic

Comment (olpipi) [reply src/cpp/src/llm_pipeline.cpp:76]: I"d like to check if the same transformation was already applied.
But I suppose the right option is to compare with PartialShape("0..1")

File: src/cpp/src/llm_pipeline.cpp (hunk 2 above)
Comment (olpipi) [reply src/cpp/src/llm_pipeline.cpp]: done

**Review by ilya-lavrenov — APPROVED** [review]

**Review by Wovchena — APPROVED** [review]

**Comment by Wovchena** [comment]
You have a check fallen

**Review by Wovchena — APPROVED** [review]
Resolve the conflict

**Pull request merged**